module/signature/random_beacon_signer_store.go

```diff
@@ -0,0 +1,109 @@
+package signature
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/onflow/flow-go/model/encoding"
+	"github.com/onflow/flow-go/module"
+	"github.com/onflow/flow-go/storage"
+)
+
+type EpochAwareRandomBeaconSignerStore struct {
+	epochLookup module.EpochLookup          // used to fetch epoch counter by view
+	keys        storage.DKGKeys             // used to fetch DKG private key by epoch
+	signers     map[uint64]module.MsgSigner // cache of signers by epoch
+}
+
+func NewEpochAwareRandomBeaconSignerStore(epochLookup module.EpochLookup, keys storage.DKGKeys) *EpochAwareRandomBeaconSignerStore {
+	return &EpochAwareRandomBeaconSignerStore{
+		epochLookup: epochLookup,
+		keys:        keys,
+		signers:     make(map[uint64]module.MsgSigner),
+	}
+}
+
+// GetSigner returns the threshold-signer for signing objects at a
+// given view. The view determines the epoch, which determines the DKG private
+// key underlying the signer.
+// It returns:
+//  - (signer, nil) if DKG was completed in the epoch of the view, signer is not nil
+//  - (nil, protocol.ErrEpochNotCommitted) if no epoch found for given view
+//  - (nil, DKGIncompleteError) if DKG was not completed in the epoch of the view
+//  - (nil, error) if there is any exception
+func (s *EpochAwareRandomBeaconSignerStore) GetSigner(view uint64) (module.MsgSigner, error) {
+	// fetching the epoch by view, if epoch is found, then DKG must have been completed
+	epoch, err := s.epochLookup.EpochForViewWithFallback(view)
+	if err != nil {
+		return nil, fmt.Errorf("could not get epoch by view %v: %w", view, err)
+	}
+
+	// when DKG has completed,
+	// 		1. if a node successfully generated the DKG key, the private key will be stored in database.
+	//    2. if a node failed to generate the DKG key, we will save a record in database to indicate this
+	//       node has no DKG key for this epoch
+	// with the epoch, we can lookup my DKG key for the epoch. There are 3 cases:
+	// 1. DKG has completed, and the private key is stored in database, and we can retrieve it (happy path)
+	// 2. DKG has completed, but we failed to generate our private key, and we marked in the database
```

Comment:
The reason for this is that DKG can generate a private key but we would judge not to use it. 
```suggestion
	// 2. DKG has completed, but we failed it, and we marked in the database
```